README.md
# Welcome to HTTP Client for JS 👋
![Version](https://img.shields.io/badge/version-1.0.0-blue.svg?cacheSeconds=2592000)

## Get Started
```npm
```




CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

### UNRELEASED

## [0.0.1] - 2023-03-30
### Added
- create http client for GET, POST, PUT, DELETE methods
- add http options for timeout and retry strategy
- add base path option 

### Changes Status Available
- Added: add new features.
- Changed: changes in features.
- Fixed: error fixing.
- Deprecated: the feature will be deprecated, without the support and deleted in new versions.
- Removed: unused features or files, deleted in this version.
- Security: in case of vulnerabilities.
- Yanked: in case of versions have to delete caused of errors or security issues. Ex: `## 0.0.5 - 2014-12-13 [YANKED]`





test/url.test.ts
import { buildURL } from '../src/url'

const url = 'http://test'
describe('URL builder test', () => {
    test('build url with param', () => {
        const response = buildURL({
            baseURL: `${url}/{1}`,
            params: ['test'],
        })
        expect(response).toEqual(`${url}/test`)
    })
    test('build url with params /test/{1}/test/{2}', () => {
        const response = buildURL({
            baseURL: `${url}/test/{1}/test/{2}`,
            params: ['foo', 'bar'],
        })
        expect(response).toEqual(`${url}/test/foo/test/bar`)
    })
    test('build url with param and query', () => {
        const queryParams = {
            foo: 'test',
            bar: 'test',
        }
        const response = buildURL({
            baseURL: `${url}/{1}`,
            params: ['test'],
            queryParams,
        })
        expect(response).toEqual(`${url}/test?foo=test&bar=test`)
    })
})


test/client.test.ts
import { HTTPClient } from '../src/client'
import nock, { type Scope } from 'nock'

const baseURL = 'https://test.com'

nock.disableNetConnect()

describe('client test', () => {
    afterEach(() => {
        nock.cleanAll()
    })
    test('GET with fail caused by timeout', async () => {
        const client = new HTTPClient({
            timeout: 1000,
        })
        const url = { baseURL }
        const scope: Scope = nock(baseURL)
            .get('/')
            .delayConnection(2000)
            .reply(500, { message: 'timeout of 1000ms exceeded' })
        const response = await client.get(url)
        expect(response.status).toEqual(500)
        expect(response.message).toEqual('timeout of 1000ms exceeded')
        scope.done()
    })
    test('GET endpoint with headers', async () => {
        const url = { baseURL }
        const scope = nock(baseURL).get('/').reply(200, 'yep')
        const client = new HTTPClient()
        const response = await client.get(url, {
            'content-type': 'application-json',
        })
        expect(response.status).toEqual(200)
        expect(response.data).toEqual('yep')
        scope.done()
    })
    test('GET endpoint with internal error', async () => {
        const url = { baseURL }
        const scope = nock(baseURL).get('/').reply(500)
        const client = new HTTPClient()
        const response = await client.get(url)
        expect(response.status).toEqual(500)
        scope.done()
    })
    test('GET endpoint with not found with retry', async () => {
        const url = { baseURL }
        const scope = nock(baseURL).get('/').reply(500).get('/').reply(404)
        const client = new HTTPClient({
            retry: 3,
            delayRetry: 1,
        })
        const response = await client.get(url)
        expect(response.status).toEqual(404)
        expect(response.data).toEqual('')
        expect(response.retryAttempt).toEqual(1)
        scope.done()
    })
    test('GET endpoint with http client basePath', async () => {
        const url = { baseURL }
        const scope = nock(baseURL).get('/').reply(200, 'yep')
        const client = new HTTPClient()
        const response = await client.get(url)
        expect(response.status).toEqual(200)
        scope.done()
    })
    test('GET endpoint with retry', async () => {
        const url = { baseURL }
        const scope = nock(baseURL)
            .get('/')
            .replyWithError('500')
            .get('/')
            .replyWithError('500')
            .get('/')
            .reply(200, 'yep')
        const client = new HTTPClient({
            retry: 5,
            delayRetry: 1,
        })

        const response = await client.get(url)
        expect(response.data).toEqual('yep')
        expect(response.retryAttempt).toEqual(2)
        scope.done()
    })
    test('POST with body not retry', async () => {
        const url = { baseURL: baseURL }
        const scope = nock(baseURL)
            .post('/', { foo: 'test' })
            .reply(200, { foo: 'test' })
        const client = new HTTPClient()
        const response = await client.post(url, { foo: 'test' }, {})
        expect(response.status).toEqual(200)
        expect(response.retryAttempt).toEqual(1)
        expect(response.data).toEqual({ foo: 'test' })
        scope.done()
    })
    test('PUT with params and body', async () => {
        const url = { baseURL: `${baseURL}/{1}`, params: ['test'] }
        const client = new HTTPClient()
        const scope = nock(baseURL)
            .put('/test', { foo: 'test' })
            .reply(201, { foo: 'test' })
        const response = await client.put(url, { foo: 'test' }, {})
        expect(response.status).toEqual(201)
        expect(response.retryAttempt).toEqual(1)
        expect(response.data).toEqual({ foo: 'test' })
        scope.done()
    })
    test('PATCH with params and body', async () => {
        const url = { baseURL: `${baseURL}/{1}`, params: ['test'] }
        const client = new HTTPClient()
        const scope = nock(baseURL)
            .patch('/test', { foo: 'test' })
            .reply(201, { foo: 'test' })
        const response = await client.patch(url, { foo: 'test' }, {})
        expect(response.status).toEqual(201)
        expect(response.retryAttempt).toEqual(1)
        expect(response.data).toEqual({ foo: 'test' })
        scope.done()
    })
    test('DELETE with exception and status 500', async () => {
        const url = { baseURL: `${baseURL}/{1}`, params: ['test'] }
        const client = new HTTPClient({
            retry: 2,
            delayRetry: 1,
        })
        const scope: Scope = nock(baseURL)
            .delete('/test')
            .reply(500)
            .delete('/test')
            .reply(500)
            .delete('/test')
            .reply(500)
        const response = await client.delete(url)
        expect(response.status).toEqual(500)
        expect(response.message).toEqual('Request failed with status code 500')
        scope.done()
    })
})


src/url.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildURL = void 0;
var build_url_ts_1 = require("build-url-ts");
var regExp = /{\d+}/g; // identify structures -> {1}
function buildURL(url) {
    var baseURL = url.baseURL, params = url.params, queryParams = url.queryParams;
    params = params != null ? params : [];
    queryParams = queryParams != null ? queryParams : undefined;
    var baseURLWithParams = buildBaseURLWithParams(baseURL, params);
    return (0, build_url_ts_1.default)(baseURLWithParams, { queryParams: queryParams });
}
exports.buildURL = buildURL;
function buildBaseURLWithParams(base, params) {
    var _a;
    var replacers = (_a = base.match(regExp)) === null || _a === void 0 ? void 0 : _a.map(function (r) { return r.toString(); });
    if (replacers === undefined) {
        return base;
    }
    params.forEach(function (value, key) {
        base = base.replace(replacers[key], value);
    });
    return base;
}


src/url.ts
import buildUrl from 'build-url-ts'

const regExp = /{\d+}/g // identify structures -> {1}

export interface clientURL {
    baseURL: string
    params?: string[]
    queryParams?: Record<string, string>
}

export function buildURL(url: clientURL): string {
    let { baseURL, params, queryParams } = url
    params = params != null ? params : []
    queryParams = queryParams != null ? queryParams : undefined
    const baseURLWithParams = buildBaseURLWithParams(baseURL, params)
    return buildUrl(baseURLWithParams, { queryParams })
}
function buildBaseURLWithParams(base: string, params: string[]): string {
    const replacers = base.match(regExp)?.map((r) => r.toString())
    if (replacers === undefined) {
        return base
    }
    params.forEach((value: string, key: number) => {
        base = base.replace(replacers[key], value)
    })
    return base
}


src/client.ts
import axios, {
    type AxiosError,
    type AxiosResponse,
    type AxiosStatic,
    type InternalAxiosRequestConfig,
} from 'axios'
import * as rax from 'retry-axios'
import { buildURL, type clientURL } from './url'

interface HTTPResponse<T> {
    status: number
    data?: T
    retryAttempt?: number
    message?: string
}
interface HTTPRequestOptions {
    timeout?: number
    retry?: number
    delayRetry?: number
}
export class HTTPClient {
    private readonly axios: AxiosStatic
    constructor(options?: HTTPRequestOptions) {
        this.axios = axios
        this.axios.defaults.raxConfig = {
            retry: options?.retry != null ? options.retry : 0,
            retryDelay: options?.delayRetry != null ? options?.delayRetry : 0,
        }
        rax.attach(this.axios)
        this.axios.interceptors.request.use(
            (request: InternalAxiosRequestConfig) => {
                request.validateStatus = (status: number) => {
                    return status < 300 || status === 404
                }
                request.timeout = options?.timeout != null ? options.timeout : 0
                return request
            }
        )
    }

    async get<T>(
        url: clientURL,
        headers?: Record<string, string>
    ): Promise<HTTPResponse<T>> {
        try {
            const builtURL = buildURL(url)
            const res = await this.axios.get(builtURL, {
                headers,
            })
            return this.resolveResponse(res)
        } catch (e) {
            return this.rejectResponse(e as AxiosError)
        }
    }

    async post<T, V>(
        url: clientURL,
        body: T,
        headers?: Record<string, string>
    ): Promise<HTTPResponse<V>> {
        try {
            const builtURL = buildURL(url)
            const res = await this.axios.post(builtURL, body, {
                headers,
            })
            return this.resolveResponse(res)
        } catch (e) {
            return this.rejectResponse(e as AxiosError)
        }
    }

    async put<T, V>(
        url: clientURL,
        body: T,
        headers?: Record<string, string>
    ): Promise<HTTPResponse<V>> {
        try {
            const builtURL = buildURL(url)
            const res = await this.axios.put(builtURL, body, {
                headers,
            })
            return this.resolveResponse(res)
        } catch (e) {
            return this.rejectResponse(e as AxiosError)
        }
    }

    async patch<T, V>(
        url: clientURL,
        patch: T,
        headers?: Record<string, string>
    ): Promise<HTTPResponse<V>> {
        try {
            const builtURL = buildURL(url)
            const res = await this.axios.patch(builtURL, patch, {
                headers,
            })
            return this.resolveResponse(res)
        } catch (e) {
            return this.rejectResponse(e as AxiosError)
        }
    }

    async delete<T>(
        url: clientURL,
        headers?: Record<string, string>
    ): Promise<HTTPResponse<T>> {
        try {
            const builtURL = buildURL(url)
            const res = await this.axios.delete(builtURL, {
                headers,
            })
            return this.resolveResponse(res)
        } catch (e) {
            return this.rejectResponse(e as AxiosError)
        }
    }

    private getCurrentRetryAttempt(
        config: InternalAxiosRequestConfig | undefined
    ): number {
        return config?.raxConfig?.currentRetryAttempt !== undefined
            ? config.raxConfig.currentRetryAttempt
            : 1
    }

    private resolveResponse(res: AxiosResponse): HTTPResponse<any> {
        const { config, status, data } = res
        const currentRetryAttempt = this.getCurrentRetryAttempt(config)
        return {
            status,
            data,
            retryAttempt: currentRetryAttempt,
        } satisfies HTTPResponse<any>
    }

    private rejectResponse(error: AxiosError): HTTPResponse<never> {
        const { response, message } = error
        const status = response != null ? response.status : 500
        return {
            status,
            message,
        } satisfies HTTPResponse<never>
    }
}


examples/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var client_1 = require("../src/client");
var getResourceFromAPI = function () { return __awaiter(void 0, void 0, void 0, function () {
    var url, client;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                url = 'https://pokeapi.co/api/v2/pokemon/ditto';
                client = new client_1.HTTPClient();
                return [4 /*yield*/, client.get({
                        baseURL: url,
                    })];
            case 1: return [2 /*return*/, _a.sent()];
        }
    });
}); };
var resource = getResourceFromAPI();
console.log(resource);


src/client.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HTTPClient = void 0;
var axios_1 = require("axios");
var rax = require("retry-axios");
var url_1 = require("./url");
var HTTPClient = /** @class */ (function () {
    function HTTPClient(options) {
        this.axios = axios_1.default;
        this.axios.defaults.raxConfig = {
            retry: (options === null || options === void 0 ? void 0 : options.retry) != null ? options.retry : 0,
            retryDelay: (options === null || options === void 0 ? void 0 : options.delayRetry) != null ? options === null || options === void 0 ? void 0 : options.delayRetry : 0,
        };
        rax.attach(this.axios);
        this.axios.interceptors.request.use(function (request) {
            request.validateStatus = function (status) {
                return status < 300 || status === 404;
            };
            request.timeout = (options === null || options === void 0 ? void 0 : options.timeout) != null ? options.timeout : 0;
            return request;
        });
    }
    HTTPClient.prototype.get = function (url, headers) {
        return __awaiter(this, void 0, void 0, function () {
            var builtURL, res, e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        builtURL = (0, url_1.buildURL)(url);
                        return [4 /*yield*/, this.axios.get(builtURL, {
                                headers: headers,
                            })];
                    case 1:
                        res = _a.sent();
                        return [2 /*return*/, this.resolveResponse(res)];
                    case 2:
                        e_1 = _a.sent();
                        return [2 /*return*/, this.rejectResponse(e_1)];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    HTTPClient.prototype.post = function (url, body, headers) {
        return __awaiter(this, void 0, void 0, function () {
            var builtURL, res, e_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        builtURL = (0, url_1.buildURL)(url);
                        return [4 /*yield*/, this.axios.post(builtURL, body, {
                                headers: headers,
                            })];
                    case 1:
                        res = _a.sent();
                        return [2 /*return*/, this.resolveResponse(res)];
                    case 2:
                        e_2 = _a.sent();
                        return [2 /*return*/, this.rejectResponse(e_2)];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    HTTPClient.prototype.put = function (url, body, headers) {
        return __awaiter(this, void 0, void 0, function () {
            var builtURL, res, e_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        builtURL = (0, url_1.buildURL)(url);
                        return [4 /*yield*/, this.axios.put(builtURL, body, {
                                headers: headers,
                            })];
                    case 1:
                        res = _a.sent();
                        return [2 /*return*/, this.resolveResponse(res)];
                    case 2:
                        e_3 = _a.sent();
                        return [2 /*return*/, this.rejectResponse(e_3)];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    HTTPClient.prototype.patch = function (url, patch, headers) {
        return __awaiter(this, void 0, void 0, function () {
            var builtURL, res, e_4;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        builtURL = (0, url_1.buildURL)(url);
                        return [4 /*yield*/, this.axios.patch(builtURL, patch, {
                                headers: headers,
                            })];
                    case 1:
                        res = _a.sent();
                        return [2 /*return*/, this.resolveResponse(res)];
                    case 2:
                        e_4 = _a.sent();
                        return [2 /*return*/, this.rejectResponse(e_4)];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    HTTPClient.prototype.delete = function (url, headers) {
        return __awaiter(this, void 0, void 0, function () {
            var builtURL, res, e_5;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        builtURL = (0, url_1.buildURL)(url);
                        return [4 /*yield*/, this.axios.delete(builtURL, {
                                headers: headers,
                            })];
                    case 1:
                        res = _a.sent();
                        return [2 /*return*/, this.resolveResponse(res)];
                    case 2:
                        e_5 = _a.sent();
                        return [2 /*return*/, this.rejectResponse(e_5)];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    HTTPClient.prototype.getCurrentRetryAttempt = function (config) {
        var _a;
        return ((_a = config === null || config === void 0 ? void 0 : config.raxConfig) === null || _a === void 0 ? void 0 : _a.currentRetryAttempt) !== undefined
            ? config.raxConfig.currentRetryAttempt
            : 1;
    };
    HTTPClient.prototype.resolveResponse = function (res) {
        var config = res.config, status = res.status, data = res.data;
        var currentRetryAttempt = this.getCurrentRetryAttempt(config);
        return {
            status: status,
            data: data,
            retryAttempt: currentRetryAttempt,
        };
    };
    HTTPClient.prototype.rejectResponse = function (error) {
        var response = error.response, message = error.message;
        var status = response != null ? response.status : 500;
        return {
            status: status,
            message: message,
        };
    };
    return HTTPClient;
}());
exports.HTTPClient = HTTPClient;
